modules/repository-url/src/main/java/org/elasticsearch/common/blobstore/url/http/RetryingHttpInputStream.java

```diff
@@ -0,0 +1,295 @@
+        if (retryCount >= maxRetries || e instanceof NoSuchFileException) {
+            logger.debug(new ParameterizedMessage("failed reading [{}] at offset [{}], retry [{}] of [{}], giving up",
+                blobURI, start + totalBytesRead, retryCount, maxRetries), e);
+            throw addSuppressedFailures(e);
+        }
+
+        logger.debug(new ParameterizedMessage("failed reading [{}] at offset [{}], retry [{}] of [{}], retrying",
+            blobURI, start + totalBytesRead, retryCount, maxRetries), e);
+
+        retryCount += 1;
+        accumulateFailure(e);
+
+        maybeAbort(delegate);
+        IOUtils.closeWhileHandlingException(delegate);
+        delegate = null;
+    }
+
+    /**
+     * Since we're using pooled http connections if we want to cancel an on-going request,
+     * we should remove that connection from the connection pool since it cannot be reused.
+     */
+    void maybeAbort(HttpResponseInputStream inputStream) {
+        if (eof || inputStream == null) {
+            return;
+        }
+
+        try {
+            if (start + totalBytesRead < currentStreamLastOffset) {
+                inputStream.abort();
+            }
+        } catch (Exception e) {
+            logger.warn("Failed to abort stream before closing", e);
+        }
+    }
+
+    private void accumulateFailure(Exception e) {
+        if (failures == null) {
+            failures = new ArrayList<>(MAX_SUPPRESSED_EXCEPTIONS);
+        }
+        if (failures.size() < MAX_SUPPRESSED_EXCEPTIONS) {
+            failures.add(e);
+        }
+    }
+
+    private IOException addSuppressedFailures(IOException e) {
+        if (failures == null) {
+            return e;
+        }
+        for (Exception failure : failures) {
+            e.addSuppressed(failure);
+        }
+        return e;
+    }
+
+    private HttpResponseInputStream openInputStream() throws IOException {
+        try {
+            return AccessController.doPrivileged((PrivilegedExceptionAction<HttpResponseInputStream>) () -> {
+                final Map<String, String> headers = new HashMap<>(1);
+
+                if (isRangeRead()) {
+                    headers.put("Range", getBytesRange(Math.addExact(start, totalBytesRead), end));
+                }
+
+                try {
+                    final URLHttpClient.HttpResponse response = httpClient.get(blobURI, headers);
+                    final int statusCode = response.getStatusCode();
+
+                    if (statusCode != RestStatus.OK.getStatus() && statusCode != RestStatus.PARTIAL_CONTENT.getStatus()) {
+                        final HttpResponseInputStream inputStream = response.getInputStream();
+                        IOUtils.closeWhileHandlingException(inputStream);
+                        IOUtils.closeWhileHandlingException(response);
+                        throw new IOException(getErrorMessage("The server returned an invalid status code :" + statusCode));
```

Comment:
Can we return more than the status code here? The server might tell us a more precise reason why it failed

Reply:
Addressed in https://github.com/elastic/elasticsearch/pull/69521/commits/ba7daf32a08bc186b39a0fbd1f689f3bff2afe6c